[pkg/ottl] Add OTTL context inferrer utility

edmocosta opened the pull request:
<!--Ex. Fixing a bug - Describe the bug and how this fixes the issue.
Ex. Adding a feature - Explain what this achieves.-->
#### Description

This PR is part of https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/29017, and adds the context inferrer utility into the `ottl` package. This utility will be used by components to infer the context it should use to parse some statements, based only on its `ottl.path`'s context prefixes.

This PR delivers a priority-based `ottl.ContextInferrer` implementation, which chooses the context based on a priority list passed on the `ottl.NewPriorityContextInferrer` function. It also exposes the  `ottl.DefaultPriorityContextInferrer` function, which creates the same context inferrer type, but using a pre-defined context priorities.

For example, given the context priority  `["spanevent", "span", "resource"]`, and inferring the context from the statement
`set(span.foo, resource.value) where spanevent.bar == true`, would result in the `spanevent` context, as it has the highest priority.

For more details on how it will be used, please have a look at the complete [draft](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/35050) implementation.

**To discuss**

_Do we really need to support the NewPriorityContextInferrer `ignoreUnknownContext` argument?_

I would say yes, considering the inferrer could be used for any purpose, having the capability of ignoring unknown/non-prioritised contexts would be helpful and wouldn't require users to check if the `Infer(..)` function returned an unexpected value.

<!-- Issue number (e.g. #1234) or full URL to issue, if applicable. -->
#### Link to tracking issue
https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/29017

<!--Describe what testing was performed and which tests were added.-->
#### Testing
Unit tests

<!--Describe the documentation added.-->
#### Documentation
No changes

edmocosta committed:
Add OTTL context inferrer utility to resolve statements context based on its paths

TylerHelmuth commented on pkg/ottl/context_inferrer.go:
```diff
@@ -0,0 +1,85 @@
+// If unknown/non-prioritized contexts are found on the statements, they can be either ignored
+// or considered when no other prioritized context is found. To skip unknown contexts, the
+// ignoreUnknownContext argument must be set to false.
+func NewPriorityContextInferrer(contextsPriority []string, ignoreUnknownContext bool) ContextInferrer {
```
My gut is saying that we'll come up with more options for this concept so I think it would be good to switch `ignoreUnknownContext bool` to the Options pattern if we keep it.

TylerHelmuth commented on pkg/ottl/context_inferrer.go:
```diff
@@ -0,0 +1,85 @@
+	}
+	return &priorityContextInferrer{
+		contextPriority:      contextPriority,
+		ignoreUnknownContext: ignoreUnknownContext,
```
Erroring on unknown contexts definitely feels like an acceptable outcome. This is a startup situation, so being able to identify a statement like `set(spna.attributes["test"], "pass")` as a typo in `spna` feels really useful.

edmocosta replied:
The current implementation is using this flag to [skip](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/35721/files#diff-ea1425b770cca2f16c5ecf77b75ed38b7dfeceb03fbc8d5e3cd1bd413ad506d8R45) non-prioritised contexts, it doesn't return an error when it founds those contexts, it either ignores it, or use it with the lowest possible priority. The responsibility of validating either the unknown context is valid or not is delegated to the [parser collection](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/35050/files#diff-7efe3affb096a0fa16edab4c8e48d665bd120c938103b778ad778b15c5f04ce6R190), which is the one that holds all supported contexts. 

The `contextPriority` here acts more like a list of well-defined priorities, and not like a list of valid contexts. One advantage on this approach is that it only needs to know the priorities, and not all (and only) the supported contexts.

TylerHelmuth replied:
In that case I think we can probably start with implementing either the skip or "lowest" behavior and adding configurability later if we need it. Whether `spna` is ignored or set as the lowest context, I would want the parser collection (or something) to be able to tell the user that `spna` is an unsupported context.

TylerHelmuth commented on pkg/ottl/context_inferrer.go:
```diff
@@ -0,0 +1,85 @@
+)
+
+// ContextInferrer is an interface used to infer the OTTL context from statements paths.
+type ContextInferrer interface {
```
Looking at the draft PR I don't believe this needs to be exported, which would be great.

edmocosta replied:
Yes, it doesn't necessarily needs to be public, my initial idea was exposing it as a `ottl.ParserCollection` [option](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/35050/files#diff-7efe3affb096a0fa16edab4c8e48d665bd120c938103b778ad778b15c5f04ce6R156), so it could support different implementations. 

I'm still have mixing feelings about that, considering we expose the `ottl.Parser[K]`, and that the new `ottl.ParserCollection` (draft) wouldn't be coupled to the existing OTTL contexts, would having a custom context inferrer be useful for something? WDYT?

Anyway, I think you're right, it's probably better not exporting it now and open it in the future if needed.

TylerHelmuth commented on pkg/ottl/context_inferrer.go:
```diff
@@ -0,0 +1,85 @@
+// ContextInferrer is an interface used to infer the OTTL context from statements paths.
+type ContextInferrer interface {
+	// Infer returns the OTTL context inferred from the given statements paths.
+	Infer(statements []string) (string, error)
```
same with this method

TylerHelmuth commented on pkg/ottl/context_inferrer.go:
```diff
@@ -0,0 +1,85 @@
+	var inferredContext string
+	var inferredContextPriority int
+
+	for _, statement := range statements {
+		parsed, err := parseStatement(statement)
+		if err != nil {
+			return inferredContext, err
+		}
+
+		for _, p := range getParsedStatementPaths(parsed) {
+			pathContextPriority, ok := s.contextPriority[p.Context]
+			if !ok {
+				if s.ignoreUnknownContext {
+					continue
+				}
+
+				// Lowest priority
+				pathContextPriority = math.MaxInt
+			}
+
+			if inferredContext == "" || pathContextPriority < inferredContextPriority {
+				inferredContext = p.Context
+				inferredContextPriority = pathContextPriority
+			}
+		}
+	}
```
I really like how this turned out.

edmocosta committed:
Change to not expose the context inferrer and removed the ignoreUnknownContext option

edmocosta commented:
Hi @TylerHelmuth! I've addressed your suggestions to under expose the context inferrer impl (https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/35721/commits/c0a8b672e62f0b444d9a137b2d48806e0f628e72). In addition to that, I've also removed the `ignoreUnknownContext` option, as we won't need that anymore. All internal usages on the draft are setting that flag to `false`, and considering we won't be exposing the context inferrer anymore, it would be a dead code. Thanks!

edmocosta committed:
Merge branch 'open-telemetry:main' into ottl-add-context-inferrer-utility

TylerHelmuth commented on .chloggen/ottl-add-context-inferrer-utility.yaml:
Since everything here is now unexported and not used yet we don't need a changelog

edmocosta replied:
Done! 👍

edmocosta committed:
Delete unneeded change log

edmocosta committed:
Merge remote-tracking branch 'origin/ottl-add-context-inferrer-utility' into ottl-add-context-inferrer-utility
